dasharo-stability/capsule-update.robot

```diff
@@ -0,0 +1,96 @@
+Library             Process
+Library             String
+Library             Telnet    timeout=20 seconds    connection_timeout=120 seconds
+Library             SSHLibrary    timeout=90 seconds
+Library             RequestsLibrary
+# TODO: maybe have a single file to include if we need to include the same
+# stuff in all test cases
+Resource            ../variables.robot
+Resource            ../keywords.robot
+Resource            ../keys.robot
+
+# TODO:
+# - document which setup/teardown keywords to use and what are they doing
+# - go threough them and make sure they are doing what the name suggest (not
+# exactly the case right now)
+Suite Setup         Run Keywords
+...                     Prepare Test Suite
+...                     AND
+...                     Skip If    not ${CAPSULE_UPDATE_SUPPORT}    Capsule Update not supported
+...                     AND
+...                     Flash Firmware If Not QEMU
+Suite Teardown      Run Keywords
+...                     Flash Firmware If Not QEMU
+...                     AND
+...                     Log Out And Close Connection
+
+
+*** Test Cases ***
+CUP001.001 Capsule Update (Ubuntu)
+    [Documentation]    This test aims to verify if Capsule Update process works.
+    Power On
+    Boot System Or From Connected Disk    ubuntu
+    Login To Linux
+    Switch To Root User
+
+    Set Prompt For Terminal    root@${UBUNTU_HOSTNAME}:/home#
+    Execute Command In Terminal    cd ..
+    Set Prompt For Terminal    root@${UBUNTU_HOSTNAME}:/#
+    Execute Command In Terminal    cd ..
+    Execute Command In Terminal    rm -r capsule_testing
+    Execute Command In Terminal    mkdir capsule_testing
+    Execute Command In Terminal    chmod 777 capsule_testing
+
+    Send File To DUT    ./dasharo-stability/capsule-update-files/CapsuleApp.efi    /capsule_testing/CapsuleApp.efi
+    Send File To DUT    ${CAPSULE_FW_FILE}    /capsule_testing/dasharo.cap
+
+    Power On
+    ${boot_menu}=    Enter Boot Menu Tianocore And Return Construction
+    Enter Submenu From Snapshot    ${boot_menu}    UEFI Shell
+    Read From Terminal Until    Shell>
+
+    ${out}=    Execute UEFI Shell Command    smbiosview -t 0
+    ${original_bios_version}=    Extract BIOS Version    ${out}
+    Execute UEFI Shell Command    FS0:
+    Execute UEFI Shell Command    cd capsule_testing
+    Execute UEFI Shell Command    CapsuleApp.efi dasharo.cap
+
+    ${out}=    Read From Terminal Until    seconds.)
+
+    ${digit}=    Get Key To Press    ${out}
+    Write Bare Into Terminal    ${digit}
+
+    Set DUT Response Timeout    120s
+    ${boot_menu}=    Enter Boot Menu Tianocore And Return Construction
+    Enter Submenu From Snapshot    ${boot_menu}    UEFI Shell
+    Read From Terminal Until    Shell>
+    ${out}=    Execute UEFI Shell Command    smbiosview -t 0
+    ${updated_bios_version}=    Extract BIOS Version    ${out}
+
+    Should Not Be Equal    ${original_bios_version}    ${updated_bios_version}
+
```

Comment:
Actually, let's not do this here, it will be in dedicated task under separate PR.

Reply:
@JanPrusinowski link PR where you are working on that, it can be PR to this `OS-block-booting-verification` branch.

Reply:
@krystian-hebel what is the expected output from `CapsuleApp.efi -S`. As on Qemu output shows:
![image](https://github.com/user-attachments/assets/5aa63f4c-a75a-4522-99c1-5bb5bdc60779)
and on MSI there is no output whatsoever:
![image](https://github.com/user-attachments/assets/289242cf-84cb-41a8-846d-45512e107c5c)

At this point I am not sure as how in anyway at this point it can be utilized in any test and what the goal is.

Reply:
This command prints status of the last capsule update attempt.  Because updating system firmware empties variable storage, this command is useful only after a failed capsule update to check the reason of the failure.

If capsule updates are enabled and there were no update attempts, `CapsuleMax - CapsuleFFFF` message is normal output.  Otherwise, you get something like:
```
CapsuleMax - CapsuleFFFF
CapsuleLast - Capsule0000
CapsuleName: Capsule0000
  Capsule Guid: 6DCBD5ED-E82D-4C44-BDA1-7194199AD92A
  Capsule ProcessedTime: 08/11/2024  21:34
  Capsule Status: Security Violation
  Capsule FMP Version: 0x1
  Capsule FMP PayloadIndex: 0x0
  Capsule FMP UpdateImageIndex: 0x1
  Capsule FMP UpdateImageTypeId: 00112233-4455-6677-8899-AABBCCDDEEFF
  Capsule FMP CapsuleFileName: ""
  Capsule FMP CapsuleTarget: ""
```
Each `CapsuleXXXX` is a variable describing a result of an update attempt.

The goal is to see what `Capsule Status:` contains after an error.  It's `Security Violation` for invalid certificate, `Not Ready` for invalid GUID (because GUID doesn't correspond to any firmware) and possibly something else for anti-rollback case.

However, these variables are set only if you do 2 reboots, so if reboots are avoided, this command isn't useful.  If EDK is updated to reject capsules right away, then maybe output of `CapsuleApp.efi dasharo.cap` could be checked instead.

Reply:
@JanPrusinowski is it clear what changes you need to make? Please confirm. This task is your highest priority, so let’s ensure there’s no delay in communication.

Reply:
I'm on it.

Reply:
I have added checks using `CapsuleApp.efi -S` in: 04ae61bc0a79a4645b2cce9c96f2ca8d6f9fd8ea

Reply:
I have made changes suggested by Sergii. They can be found in the PR.

Reply:
> I have made changes suggested by Sergii. They can be found in the PR.

@JanPrusinowskicould you please link those specific changes, so the reviewer can verify if they have been made correctly?

Reply:
Changes can be found in: 04ae61bc0a79a4645b2cce9c96f2ca8d6f9fd8ea and: 59684c89fcf0d575ad079d2e7682798793b03a8b

Changes adressed checks with 'CapsuleApp.efi -S' after performing update.